File: ManagedSteam/Implementations/SteamController.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Runtime.InteropServices;

using ManagedSteam.CallbackStructures;
using ManagedSteam.SteamTypes;
using ManagedSteam.Utility;

namespace ManagedSteam.Implementations
{
    class SteamController : SteamService, ISteamController
    {
        public SteamController()
        {

        }

        internal override void CheckIfUsableInternal()
        {

        }

        internal override void InvokeEvents()
        {

        }

        internal override void ReleaseManagedResources()
        {

        }

        public bool Init(string absolutPathToControllerConfigVDF)
        {
            CheckIfUsable();

            return NativeMethods.SteamController_Init( absolutPathToControllerConfigVDF );
        }

        public bool Shutdown()
        {
            CheckIfUsable();

            return NativeMethods.SteamController_Shutdown();
        }

        public void RunFrame()
        {
            CheckIfUsable();

            NativeMethods.SteamController_RunFrame();
        }

        public bool GetControllerState(uint controllerIndex, out SteamControllerState state)
        {
            CheckIfUsable();

            using (NativeBuffer bufferKey = new NativeBuffer(Marshal.SizeOf(typeof(SteamControllerState))))
            {
                bool result = NativeMethods.SteamController_GetControllerState(controllerIndex, bufferKey.UnmanagedMemory);
                state = NativeHelpers.ConvertStruct<SteamControllerState>(bufferKey.UnmanagedMemory, bufferKey.UnmanagedSize);

                return result;
            }
        }

        public SteamControllerGetControllerStateResult GetControllerState(uint controllerIndex)
        {
            CheckIfUsable();

            SteamControllerGetControllerStateResult result;

            using (NativeBuffer bufferKey = new NativeBuffer(Marshal.SizeOf(typeof(SteamControllerState))))
            {
                result.Result = NativeMethods.SteamController_GetControllerState(controllerIndex, bufferKey.UnmanagedMemory);
                result.State = NativeHelpers.ConvertStruct<SteamControllerState>(bufferKey.UnmanagedMemory, bufferKey.UnmanagedSize);

                return result;
            }
        }

        public void TriggerHapticPulse(uint controllerIndex, SteamControllerPad targetPad, ushort durationMicroSec)
        {
            NativeMethods.SteamController_TriggerHapticPulse( controllerIndex, (int)targetPad, durationMicroSec );
        }

        public void SetOverrideMode(string mode)
        {
            NativeMethods.SteamController_SetOverrideMode( mode );
        }
    }
}

File: SteamAPIWrap/SteamController.cpp
#include "Precompiled.hpp"
#include "SteamController.hpp"

#include "MemoryHelpers.h"

using namespace SteamAPIWrap;

namespace SteamAPIWrap
{
	template <> CSteamController * CSingleton<CSteamController>::instance = nullptr;

	CSteamController::CSteamController()
		: steamcontroller(nullptr)
	{
	}

	void CSteamController::SetSteamInterface(ISteamController *steamcontroller)
	{
		this->steamcontroller = steamcontroller;
	}

	bool CSteamController::Init( PConstantString absolutPathToControllerConfigVDF )
	{
		return steamcontroller->Init( absolutPathToControllerConfigVDF );
	}

	bool CSteamController::Shutdown()
	{
		return steamcontroller->Shutdown();
	}

	void CSteamController::RunFrame()
	{
		steamcontroller->RunFrame();
	}

	bool CSteamController::GetControllerState( u32 controllerIndex, PDataPointer state )
	{
		return steamcontroller->GetControllerState( controllerIndex, reinterpret_cast<SteamControllerState_t *>(state));
	}

	void CSteamController::TriggerHapticPulse( u32 controllerIndex, Enum targetPad, u16 durationMicroSec )
	{
		steamcontroller->TriggerHapticPulse(controllerIndex, static_cast<ESteamControllerPad>(targetPad), durationMicroSec);
	}

	void CSteamController::SetOverrideMode( PConstantString mode )
	{
		steamcontroller->SetOverrideMode(mode);
	}
};

ManagedSteam_API bool SteamController_Init( PConstantString absolutPathToControllerConfigVDF )
{
	return CSteamController::Instance().Init( absolutPathToControllerConfigVDF );
}

ManagedSteam_API bool SteamController_Shutdown()
{
	return CSteamController::Instance().Shutdown();
}

ManagedSteam_API void SteamController_RunFrame()
{
	CSteamController::Instance().RunFrame();
}

ManagedSteam_API bool SteamController_GetControllerState( u32 controllerIndex, PDataPointer state )
{
	return CSteamController::Instance().GetControllerState( controllerIndex, state );
}

ManagedSteam_API void SteamController_TriggerHapticPulse( u32 controllerIndex, Enum targetPad, u16 durationMicroSec )
{
	CSteamController::Instance().TriggerHapticPulse( controllerIndex, targetPad, durationMicroSec );
}

ManagedSteam_API void SteamController_SetOverrideMode( PConstantString mode )
{
	CSteamController::Instance().SetOverrideMode( mode );
}